components/script/dom/shadowroot.rs

```diff
@@ -326,9 +326,24 @@ impl VirtualMethods for ShadowRoot {
         let shadow_root = self.upcast::<Node>();
 
         shadow_root.set_flag(NodeFlags::IS_CONNECTED, context.tree_connected);
-        for node in shadow_root.children() {
+
+        for node in shadow_root.traverse_preorder(ShadowIncluding::No) {
             node.set_flag(NodeFlags::IS_CONNECTED, context.tree_connected);
-            node.bind_to_tree(context);
+
+            if context.tree_is_in_a_shadow_tree {
+                if let Some(shadow_root) = node.containing_shadow_root() {
+                    node.set_containing_shadow_root(Some(&*shadow_root));
+                }
+                debug_assert!(node.containing_shadow_root().is_some());
+            }
+
+            // Out-of-document elements never have the descendants flag set
+            debug_assert!(!node.get_flag(NodeFlags::HAS_DIRTY_DESCENDANTS));
+            vtable_for(&node).bind_to_tree(&BindContext {
+                tree_connected: context.tree_connected,
+                tree_is_in_a_document_tree: context.tree_is_in_a_document_tree,
+                tree_is_in_a_shadow_tree: context.tree_is_in_a_shadow_tree,
```

Comment:
Since children of a shadow root are always in a shadow tree and never in a document tree, you can change this to
```diff
                tree_is_in_a_document_tree: false,
                tree_is_in_a_shadow_tree: true,
```
The only flag that we care about here is whether the tree is connected (meaning it's shadow inclusive ancestor is a document)